### PR TITLE
Svelte: Perforce UI elements refactor

### PR DESCRIPTION
**Pull request by jasonhawkharris**
## UI Updates for Perforce Depots and Git Repos

Fixes SRCH-530

**NOTE: This PR is a refactor of an earlier [PR](https://github.com/sourcegraph/sourcegraph/pull/64014) that was reverted. For that reason, the PR description is largely the same.**

This PR introduces changes to the UI to differentiate between Perforce Depots and Git repositories. Below are the key changes included in this commit:

### 1. Dynamic Top-Level Navigation

**For Perforce Depots:**

Screenshot 2024-07-31 at 10 10 37 AM

**For Git Repos:**

Screenshot 2024-07-31 at 10 10 14 AM

### 2. Tabs on Revision Picker

**For Perforce Depots:**

Since we only need one tab for changelists, no tabs are shown.

Screenshot 2024-07-31 at 10 20 24 AM

**For Git Repos:**

We have tabs for Branches, Tags, and Commits.

Screenshot 2024-07-31 at 10 23 02 AM

### 3. Commits/Changelists Page

**For Git Repos:**

The page displays Git commits.

Screenshot 2024-07-31 at 10 26 23 AM

**For Perforce Depots:**

The page displays Perforce changelists.

Screenshot 2024-07-31 at 10 26 39 AM

### 4. Vocabulary Adjustments

- We display either Git commit SHAs or Changelist IDs based on the project type.
- For authorship, we use "submitted by" for Perforce and "committed by" for Git.
- We refer to "Commits" for Git projects and "Changelists" for Perforce projects.

**Examples:**

- **For Git Commits:**

Screenshot 2024-07-31 at 10 37 08 AM

- **For Perforce Changelists:**

Screenshot 2024-07-31 at 10 37 35 AM

### 5. URL Mapping

URLs are now structured differently based on the project type:

- **Commits Page:**
  - Git: `/[repo-name]/-/commits`
  - Perforce: `/[repo-name]/-/changelists`
  
- **Individual Item Page:**
  - Git: `/[repo-name]/-/commit/[commit-hash]`
  - Perforce: `/[depot-name]/-/changelist/[changelist-ID]`

When viewing a specific commit or changelist:
- **Git:** `/[repo-name]@[git-commit-hash]`
- **Perforce:** `/[repo-name]@changelist/[changelist-id]`

_NOTE: The value displayed in the search field will also change accordingly._

### What is left to be done?
**On repo search results, when searching a revision, we still show the git commit SHA instead of the changelist ID for perforce depots:**
Screenshot 2024-07-31 at 10 59 12 AM

I plan to make a follow-up issue for this and begin work on it immediately. It's a little trickier than the other changes because in the RepositoryMatch type, there is no value that can help us determine whether a project is a depot or a repo. We need to find another way to fetch that data. 

### Request for reviewers: 
1. Please try to break these new features and tell me what you find. I stumbled on a number of little gotchas while working on this, and I'm sure I've missed some. 

## Test plan

- Manual/Visual testing
- Adjust e2e and integration tests to obtain a passing CI
- Test directly visiting a URL versus getting there via click
- Add unit tests for new/updated helper functions

### CODE REVIEW

File: client/web-sveltekit/src/routes/[...repo=reporev]/-/changelist/[...revspec]/+page.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import { IncrementalRestoreStrategy, getGraphQLClient, infinityQuery } from '$lib/graphql'
+
+import type { PageLoad } from './$types'
+import { ChangelistPage_ChangelistQuery, ChangelistPage_DiffQuery } from './page.gql'
+
+const PAGE_SIZE = 20
+
+export const load: PageLoad = async ({ params }) => {
+    const client = getGraphQLClient()
+    const { repoName, revision } = parseRepoRevision(params.repo + '@' + params.revspec)
```
Comment (camdencheek) [inline client/web-sveltekit/src/routes/[...repo=reporev]/-/changelist/[...revspec]/+page.ts:14]: I don't think we need to use `parseRepoRevision` here. We can just use the changelist id directly from the query param

File: client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts [hunk 2]
```diff
@@ -144,6 +147,23 @@ async function loadFileView({ parent, params, url }: PageLoadEvent) {
...
     const filePath = decodeURIComponent(params.path)
+    const isPerforceDepot = await parent().then(parent => parent.isPerforceDepot)
+
+    if (isPerforceDepot) {
+        const changelistInfo = await client
+            .query(DepotChangelist, {
+                depotName: repoName,
+                revision: revision,
+            })
+            .then(result => result.data?.repository?.commit?.perforceChangelist)
```
Comment (camdencheek) [inline client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts:158]: For this, instead of making a second, blocking request every time we load a file page, can we just add changelist information to the `ResolveRepository` query that we run in the layout? That way, we can just do something like `if ((await parent()).resolvedRepository.commit.changelistID !== undefined) { redirect(...) }`

File: client/web-sveltekit/src/lib/repo/RepositoryRevPicker.svelte [hunk 3]
```diff
@@ -258,7 +296,7 @@
             }
         }
 
-        .commit-subject {
+        .changelist-subject {
```
Comment (camdencheek) [inline client/web-sveltekit/src/lib/repo/RepositoryRevPicker.svelte:299]: does renaming this break the style of commits?

File: client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts (hunk 2 above)
Comment (camdencheek) [reply client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts:158]: Alternatively, we could just do the redirect in the layout data loader so it works for every subpage that supports `repo@rev` in the URL (this is probably the better option)

Comment (fkling) [reply client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts:158]: I'll take a closer look on Monday, but I agree that that making a blocking request here at all is not great. I understand that we can't do the redirect otherwise but I wonder if we actually have to do that.
But as it is now we won't load the blob data until we've loaded the repo data.

Comment (camdencheek) [reply client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts:158]: I think if we move it to the layout, it doesn't need to block at all. We just throw a redirect when the repo data loads, but loading the file data can remain indepenent.

File: client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/+layout.svelte [hunk 4]
```diff
@@ -212,16 +212,28 @@
...
+                {#if data.isPerforceDepot}
+                    <RepositoryRevPicker
+                        display="block"
+                        repoURL={data.repoURL}
+                        revision={data.revision}
+                        commitID={data.resolvedRevision.commitID}
+                        defaultBranch={data.defaultBranch}
+                        latestChangelistID={data.latestChangelistID}
+                        getDepotChangelists={data.getDepotChangelists}
+                    />
```
Comment (fkling) [inline client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/+layout.svelte:224]: If the corresponding changes this could be
```suggestion
                    <RepositoryRevPicker
                        display="block"
                        repoURL={data.repoURL}
                        commitID={data.latestChangelistID}
                        getDepotChangelists={data.getDepotChangelists}
                    />
```

File: client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/tree/[...path]/+page.ts [hunk 5]
```diff
@@ -1,17 +1,38 @@
...
+            })
+            .then(result => result.data?.repository?.commit?.perforceChangelist)
+
+        if (changelistInfo?.cid) {
+            const redirectURL = new URL(url)
+
+            redirectURL.pathname = `${repoName}@changelist/${changelistInfo?.cid}/-/tree/${filePath}`
+            redirect(301, redirectURL)
+        }
+    }
```
Comment (fkling) [inline client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/tree/[...path]/+page.ts:35]: Same as for the file page.

File: client/web-sveltekit/src/routes/[...repo=reporev]/+layout.ts [hunk 6]
```diff
@@ -34,6 +36,13 @@ export const load: LayoutLoad = async ({ params, url, depends }) => {
         url,
     })
 
+    const latestChangelistID = await client
+        .query(DepotChangelist, {
+            depotName: repoName,
+            revision: revision,
+        })
+        .then(res => res.data?.repository?.commit?.perforceChangelist?.cid)
```
Comment (fkling) [inline client/web-sveltekit/src/routes/[...repo=reporev]/+layout.ts:44]: I looks like we are already fetching this information as part of `resolvedRepository`. Maybe we just have to adjust the query and the caching mechanism (in `resolveRepoRevision`). We don't have to make a separate query.

Also I think the changelist/revisionID should not be returned here but should be validated in the `(validrev)/+layout.ts` loader.

**Review by fkling — COMMENTED** [review]
I still don't know how to feel about context, but I wonder whether we should make `isPerforceDepot` available via context instead of passing around props. Curious was @camdencheek thinks.

File: client/web-sveltekit/src/lib/repo/RepositoryRevPickerItem.svelte [hunk 7]
```diff
@@ -68,12 +68,6 @@
                 color: var(--icon-color);
             }
         }
-
-        // Branch name badge
-        :global([data-badge]) {
-            overflow: hidden;
-            text-overflow: ellipsis;
-        }
```
[the remark this answers is not in the document]
Comment (jasonhawkharris) [reply client/web-sveltekit/src/lib/repo/RepositoryRevPickerItem.svelte:76]: This was causing perforce changelist IDs to cut off inconsistently. When I removed it, it fixed the issue, without affecting the abbreviatedOIDs when the context is changed to git. I tried to see what else this changed but this worked well enough and I didn't have time to really dive in so I left it in. 

Do you see any issue with removing this?

[the remark this answers is not in the document]
Comment (fkling) [reply client/web-sveltekit/src/lib/repo/RepositoryRevPickerItem.svelte:76]: > When I removed it, it fixed the issue, without affecting the abbreviatedOIDs when the context is changed to git.

What about really long branch names?

File: client/web-sveltekit/src/lib/repo/RepositoryRevPicker.svelte [hunk 8]
```diff
@@ -258,6 +310,7 @@
             }
         }
 
+        .changelist-subject,
```
Comment (fkling) [inline client/web-sveltekit/src/lib/repo/RepositoryRevPicker.svelte:313]: Just use a single CSS class, `.subject`, for both elements.

File: client/web-sveltekit/src/lib/Commit.svelte
[the remark this answers is not in the document]
Comment (jasonhawkharris) [reply client/web-sveltekit/src/lib/Commit.svelte:1]: I can't see what you're referring to, exactly.

File: client/web-sveltekit/src/lib/repo/RepositoryRevPickerItem.svelte (hunk 7 above)
[the remark this answers is not in the document]
Comment (jasonhawkharris) [reply client/web-sveltekit/src/lib/repo/RepositoryRevPickerItem.svelte:76]: Yeah. It cuts them off. I didn't check for very long branch names before.
Screenshot 2024-08-12 at 6 18 08 PM

[the remark this answers is not in the document]
Comment (fkling) [reply client/web-sveltekit/src/lib/repo/RepositoryRevPickerItem.svelte:76]: Is the main problem that the space available for changelist IDs is too small? Maybe we have to adjust the width of the column for this specifically.

[the remark this answers is not in the document]
Comment (jasonhawkharris) [reply client/web-sveltekit/src/lib/repo/RepositoryRevPickerItem.svelte:76]: Ahhh... I see. You can override this. Problem solved.

File: client/web-sveltekit/src/lib/Commit.svelte
[the remark this answers is not in the document]
Comment (jasonhawkharris) [reply client/web-sveltekit/src/lib/Commit.svelte:1]: This is a band-aid for the moment. Making Commit.svelte and Changelist.svelte truly independent of each other is the first bit of follow-up work I plan to do. Are you okay merging like this if that's the case or would you feel uncomfortable doing so until this is fixed as well?

File: client/web-sveltekit/src/lib/Changelist.svelte [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    import Avatar from '$lib/Avatar.svelte'
+    import Icon from '$lib/Icon.svelte'
+    import Timestamp from '$lib/Timestamp.svelte'
+    import Tooltip from '$lib/Tooltip.svelte'
+
+    import type { Changelist } from './Changelist.gql'
+    import { isViewportMobile } from './stores'
+    import Button from './wildcard/Button.svelte'
+
+    export let changelist: Changelist
```
Comment (camdencheek) [inline client/web-sveltekit/src/lib/Changelist.svelte:13]: Live review: we should be using this fragment in the query that generates this input

File: client/web-sveltekit/src/lib/repo/RepositoryRevPickerItem.svelte [hunk 10]
```diff
@@ -93,4 +87,10 @@
         justify-content: flex-end;
         color: var(--text-muted);
     }
+
+    // Branch name badge
+    :global([data-badge]) {
+        overflow: hidden;
+        text-overflow: ellipsis;
+    }
```
Comment (fkling) [inline client/web-sveltekit/src/lib/repo/RepositoryRevPickerItem.svelte:95]: This shouldn't be here at this level, it will apply to every badge on the page. This should live inside `.title` as it previously.

File: client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/+layout.svelte (hunk 4 above)
Comment (jasonhawkharris) [reply client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/+layout.svelte]: I'll work on this next. I'd like to merge as is for now.

**Review by peterguy — APPROVED** [review]
Overall looks great! Left a couple of comments about specific things to address, but will approve to unblock; gotta get this out there so we can iterate!

File: client/web-sveltekit/src/routes/[...repo=reporev]/-/commit/[...revspec]/+page.ts [hunk 11]
```diff
@@ -19,11 +19,16 @@ export const load: PageLoad = async ({ params }) => {
...
     if (!commit) {
         error(404, 'Commit not found')
     }
 
+    if (commit.perforceChangelist !== null) {
+        const redirectURL = new URL(url)
+        redirectURL.pathname = `${params.repo}/-/changelist/${commit.perforceChangelist?.cid}`
+        redirect(301, redirectURL)
+    }
+
```
[the remark this answers is not in the document]
Comment (camdencheek) [reply client/web-sveltekit/src/routes/[...repo=reporev]/-/commit/[...revspec]/+page.ts:31]: Oh yeah, I was thinking we already had this. Let's save for a followup PR though

File: client/web-sveltekit/src/testing/integration.ts [hunk 12]
```diff
@@ -63,6 +63,7 @@ const defaultMocks: TypeMocks = {
     GitCommit: () => ({
         abbreviatedOID: faker.git.commitSha({ length: 7 }),
         subject: faker.git.commitMessage(),
+        perforceChangelist: null,
```
Comment (camdencheek) [inline client/web-sveltekit/src/testing/integration.ts:66]: @jasonhawkharris I was confused why so  many testss were failing, but it turns out this was all that was needed to fix most of them. We just needed to change the default mock so that we weren't generating a perforce changlist ID by default, which was causing us to redirect for a bunch of pages

**Review by camdencheek — APPROVED** [review]
Approved pending CI passing and Peter's suggestion about the null comparison. 

@fkling We took another pass together and I think we resolved all your concerns about the redirects. The redirects now require zero additional requests. Additionally, we updated the default branch stuff so we can just use defaultBranch directly even in perforce repos.

File: client/web-sveltekit/src/lib/repo/RepositoryRevPicker.svelte [hunk 13]
```diff
@@ -42,9 +46,10 @@
     import ButtonGroup from '$lib/wildcard/ButtonGroup.svelte'
     import CopyButton from '$lib/wildcard/CopyButton.svelte'
 
+    import { RevPickerChangelist } from '../../routes/[...repo=reporev]/layout.gql'
```
Comment (fkling) [inline client/web-sveltekit/src/lib/repo/RepositoryRevPicker.svelte:49]: If the rev picker is the only place where this query is used, then it should be defined in `RepositoryRevPicker.gql`, just like we do for commits, branches and tags, and should only be imported as `type`.

File: client/web-sveltekit/src/lib/repo/RepositoryRevPicker.svelte [hunk 14]
```diff
@@ -72,12 +83,14 @@
...
-    export let getRepositoryBranches: (query: string) => PromiseLike<RepositoryBranches>
+    export let getRepositoryTags: ((query: string) => PromiseLike<RepositoryTags>) | undefined = undefined
+    export let getRepositoryCommits: ((query: string) => PromiseLike<RepositoryCommits>) | undefined = undefined
+    export let getRepositoryBranches: ((query: string) => PromiseLike<RepositoryBranches>) | undefined = undefined
+    export let getDepotChangelists: ((query: string) => PromiseLike<DepotChangelists>) | undefined = undefined
 
     function defaultHandleSelect(revision: string) {
-        goto(replaceRevisionInURL(location.pathname + location.search + location.hash, revision))
+        const urlToReplace = location.pathname + location.search + location.hash
+        goto(replaceRevisionInURL(urlToReplace, revision))
```
Comment (fkling) [inline client/web-sveltekit/src/lib/repo/RepositoryRevPicker.svelte:93]: Unnecessary change.

File: client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts [hunk 15]
```diff
@@ -144,6 +147,23 @@ async function loadFileView({ parent, params, url }: PageLoadEvent) {
...
+
+        if (changelistInfo?.cid) {
+            const redirectURL = new URL(url)
+
+            redirectURL.pathname = `${repoName}@changelist/${changelistInfo?.cid}/-/blob/${filePath}`
+            redirect(301, redirectURL)
+        }
+    }
```
Comment (fkling) [inline client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts:166]: The changes to this file and the tree page file will still be reverted, right? If yes, the changes to the corresponding `page.gql` need to be reverted too.

File: client/web-sveltekit/src/testing/integration.ts (hunk 12 above)
Comment (jasonhawkharris) [reply client/web-sveltekit/src/testing/integration.ts:66]: Oh, awesome. Thank you so much for doing that!

File: client/web-sveltekit/src/routes/[...repo=reporev]/-/commit/[...revspec]/+page.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (jasonhawkharris) [reply client/web-sveltekit/src/routes/[...repo=reporev]/-/commit/[...revspec]/+page.ts:31]: I've created an issue for this and added it to my TODOs for this sprint.  https://linear.app/sourcegraph/issue/SRCH-530/add-perforce-ui-elements

File: client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/+layout.ts [hunk 16]
```diff
@@ -17,6 +17,16 @@ export const load: LayoutLoad = async ({ parent }) => {
         error(404, new RevisionNotFoundError(revision))
     }
 
+    const isPerforceDepot = resolvedRepository.commit?.perforceChangelist !== null
```
[the remark this answers is not in the document]
Comment (jasonhawkharris) [reply client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/+layout.ts:20]: Nice. Willd efinitely keep this in mind. Yay javascript!

File: client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts (hunk 15 above)
Comment (jasonhawkharris) [reply client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts:166]: My understanding is that these will remain where they are because they're a little different from the redirect we've added to the parent loader. They differ in that the redirect you've commented on here _makes a copy_ of the `url` object and changes the `pathname` property so that we can still use the other properties of the initial `url` object. Whereas the redirect in the parent loader does not make a copy of the URL object since it will only ever be taking a static endpoint.

@camdencheek can you check my math here?

Comment (camdencheek) [reply client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts:166]: This is the one we replaced with the redirect in the layout and can be removed. The one that is different is the redirect from `/commit` to `/changelist`, which we still want because we want to replace the changelist number in a different place in the URL.

Comment (jasonhawkharris) [reply client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.ts:166]: I'll make a note to revisit this in my follow-up work. Going to resolve this for now.
